File: ios/parent/Services/ParentAPIService.swift
import Foundation

public protocol ParentAPIServiceProtocol {
    func login(email: String, password: String) async throws -> String
    func fetchChildren(token: String) async throws -> [ParentChildModel]
    func fetchAbsences(token: String, childId: String) async throws -> [AbsenceModel]
    func justifyAbsence(token: String, childId: String, absenceId: String, reason: String) async throws -> Bool
}

public class HttpParentAPIService: ParentAPIServiceProtocol {
    private let baseURL: URL

    public init(baseURL: URL? = nil) {
        if let baseURL = baseURL {
            self.baseURL = baseURL
        } else if let url = URL(string: AppConfig.baseURL) {
            self.baseURL = url
        } else {
            self.baseURL = URL(string: "https://registro-backend-fdu2.onrender.com/api/v1")!
        }
    }

    public func login(email: String, password: String) async throws -> String {
        var cleanEmail = email.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        if cleanEmail.hasPrefix("genitorea_") {
            cleanEmail = cleanEmail.replacingOccurrences(of: "genitorea_", with: "genitore2a_")
        } else if cleanEmail.hasPrefix("genitoreb_") {
            cleanEmail = cleanEmail.replacingOccurrences(of: "genitoreb_", with: "genitore2b_")
        }

        let url = baseURL.appendingPathComponent("auth/login")
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let body = ["email": cleanEmail, "password": password]
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, response) = try await URLSession.shared.data(for: request)
        guard let httpResponse = response as? HTTPURLResponse, (200...299).contains(httpResponse.statusCode) else {
            throw NSError(domain: "ParentAPI", code: 401, userInfo: [NSLocalizedDescriptionKey: "Credenziali non valide"])
        }

        if let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] {
            let userObj = json["user"] as? [String: Any]
            let role = userObj?["role"] as? String
            guard role == "parent" else {
                throw NSError(domain: "ParentAPI", code: 403, userInfo: [NSLocalizedDescriptionKey: "Accesso non consentito: questo account non appartiene a un genitore."])
            }
            if let token = json["token"] as? String ?? json["access_token"] as? String {
                return token
            }
        }
        throw NSError(domain: "ParentAPI", code: 500, userInfo: [NSLocalizedDescriptionKey: "Token mancante"])
    }

    public func fetchChildren(token: String) async throws -> [ParentChildModel] {
        let url = baseURL.appendingPathComponent("users/me/children")
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")

        let (data, response) = try await URLSession.shared.data(for: request)
        guard let httpResponse = response as? HTTPURLResponse, (200...299).contains(httpResponse.statusCode) else {
            throw NSError(domain: "ParentAPI", code: 401, userInfo: [NSLocalizedDescriptionKey: "Non autorizzato"])
        }

        guard let jsonArray = try JSONSerialization.jsonObject(with: data) as? [[String: Any]] else {
            return []
        }

        return jsonArray.enumerated().map { index, dict in
            ParentChildModel(
                id: (dict["id"] as? String) ?? "\(index)",
                firstName: (dict["first_name"] as? String) ?? (dict["name"] as? String) ?? "Figlio",
                lastName: (dict["last_name"] as? String) ?? "",
                className: (dict["class_name"] as? String) ?? "Classe"
            )
        }
    }

    public func fetchAbsences(token: String, childId: String) async throws -> [AbsenceModel] {
        let url = baseURL.appendingPathComponent("attendance/child-attendance/\(childId)")
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")

        guard let (data, response) = try? await URLSession.shared.data(for: request),
              let httpResponse = response as? HTTPURLResponse, (200...299).contains(httpResponse.statusCode),
              let jsonArray = try? JSONSerialization.jsonObject(with: data) as? [[String: Any]] else {
            return []
        }

        return jsonArray.enumerated().map { index, dict in
            AbsenceModel(
                id: (dict["id"] as? String) ?? "\(index)",
                childId: childId,
                date: (dict["date"] as? String) ?? "",
                type: (dict["type"] as? String) ?? "Assenza",
                isJustified: (dict["is_justified"] as? Bool) ?? false,
                justificationNote: (dict["justification_note"] as? String) ?? (dict["reason"] as? String) ?? ""
            )
        }
    }

    public func justifyAbsence(token: String, childId: String, absenceId: String, reason: String) async throws -> Bool {
        let url = baseURL.appendingPathComponent("attendance/child/\(childId)/justify/\(absenceId)")
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let body = ["reason": reason]
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (_, response) = try await URLSession.shared.data(for: request)
        guard let httpResponse = response as? HTTPURLResponse, (200...299).contains(httpResponse.statusCode) else {
            return false
        }
        return true
    }
}

File: android/parent/src/main/java/it/scuola/registro/parent/network/ParentApiService.kt
package it.scuola.registro.parent.network

import it.scuola.registro.parent.data.ColloquioBooking
import it.scuola.registro.parent.data.ParentChild
import it.scuola.registro.parent.data.PendingAbsence
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject
import java.io.BufferedReader
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.net.HttpURLConnection
import java.net.URL

interface ParentApiService {
    suspend fun login(email: String, password: String): Result<String>
    suspend fun getChildren(token: String): Result<List<ParentChild>>
    suspend fun getAbsences(token: String, childId: String): Result<List<PendingAbsence>>
    suspend fun submitJustification(token: String, absenceId: String, note: String): Result<Boolean>
    suspend fun bookColloquio(token: String, colloquioId: String): Result<Boolean>
}

class HttpParentApiService(
    private val baseUrl: String = "https://registro-backend-fdu2.onrender.com/api/v1"
) : ParentApiService {

    override suspend fun login(email: String, password: String): Result<String> =
        withContext(Dispatchers.IO) {
            try {
                val url = URL("$baseUrl/auth/login")
                val conn = (url.openConnection() as HttpURLConnection).apply {
                    requestMethod = "POST"
                    setRequestProperty("Content-Type", "application/json; charset=UTF-8")
                    doOutput = true
                    doInput = true
                    connectTimeout = 10000
                    readTimeout = 10000
                }

                val payload = JSONObject().apply {
                    put("email", email)
                    put("password", password)
                }

                OutputStreamWriter(conn.outputStream).use { it.write(payload.toString()) }

                val responseCode = conn.responseCode
                if (responseCode in 200..299) {
                    val response = BufferedReader(InputStreamReader(conn.inputStream)).use { it.readText() }
                    val json = JSONObject(response)
                    val userObj = json.optJSONObject("user")
                    val role = userObj?.optString("role")
                    if (role != "parent") {
                        Result.failure(Exception("Accesso non consentito: questo account non appartiene a un genitore."))
                    } else {
                        val token = json.optString("token", json.optString("access_token"))
                        Result.success(token)
                    }
                } else {
                    val err = BufferedReader(InputStreamReader(conn.errorStream ?: conn.inputStream)).use { it.readText() }
                    Result.failure(Exception("Login fallito (HTTP $responseCode): $err"))
                }
            } catch (e: Exception) {
                Result.failure(e)
            }
        }

    override suspend fun getChildren(token: String): Result<List<ParentChild>> =
        withContext(Dispatchers.IO) {
            try {
                val url = URL("$baseUrl/parents/my-children")
                val conn = (url.openConnection() as HttpURLConnection).apply {
                    requestMethod = "GET"
                    setRequestProperty("Authorization", "Bearer $token")
                    setRequestProperty("Accept", "application/json")
                    connectTimeout = 10000
                    readTimeout = 10000
                }

                if (conn.responseCode in 200..299) {
                    val response = BufferedReader(InputStreamReader(conn.inputStream)).use { it.readText() }
                    val list = mutableListOf<ParentChild>()
                    val arr = JSONArray(response)
                    for (i in 0 until arr.length()) {
                        val obj = arr.getJSONObject(i)
                        list.add(
                            ParentChild(
                                id = obj.optString("id", i.toString()),
                                firstName = obj.optString("first_name", obj.optString("name", "Figlio")),
                                lastName = obj.optString("last_name", ""),
                                className = obj.optString("class_name", "Classe")
                            )
                        )
                    }
                    Result.success(list)
                } else {
                    Result.failure(Exception("Errore recupero figli: HTTP ${conn.responseCode}"))
                }
            } catch (e: Exception) {
                Result.failure(e)
            }
        }

    override suspend fun getAbsences(token: String, childId: String): Result<List<PendingAbsence>> =
        withContext(Dispatchers.IO) {
            try {
                val url = URL("$baseUrl/attendance/student/$childId")
                val conn = (url.openConnection() as HttpURLConnection).apply {
                    requestMethod = "GET"
                    setRequestProperty("Authorization", "Bearer $token")
                    setRequestProperty("Accept", "application/json")
                    connectTimeout = 10000
                    readTimeout = 10000
                }

                if (conn.responseCode in 200..299) {
                    val response = BufferedReader(InputStreamReader(conn.inputStream)).use { it.readText() }
                    val list = mutableListOf<PendingAbsence>()
                    val arr = JSONArray(response)
                    for (i in 0 until arr.length()) {
                        val obj = arr.getJSONObject(i)
                        list.add(
                            PendingAbsence(
                                id = obj.optString("id", i.toString()),
                                childId = childId,
                                date = obj.optString("date", ""),
                                type = obj.optString("type", "Assenza"),
                                reason = obj.optString("reason", ""),
                                isJustified = obj.optBoolean("is_justified", false)
                            )
                        )
                    }
                    Result.success(list)
                } else {
                    Result.failure(Exception("Errore recupero assenze: HTTP ${conn.responseCode}"))
                }
            } catch (e: Exception) {
                Result.failure(e)
            }
        }

    override suspend fun submitJustification(token: String, absenceId: String, note: String): Result<Boolean> =
        withContext(Dispatchers.IO) {
            try {
                val url = URL("$baseUrl/attendance/$absenceId/justify")
                val conn = (url.openConnection() as HttpURLConnection).apply {
                    requestMethod = "POST"
                    setRequestProperty("Authorization", "Bearer $token")
                    setRequestProperty("Content-Type", "application/json; charset=UTF-8")
                    doOutput = true
                    doInput = true
                    connectTimeout = 10000
                    readTimeout = 10000
                }

                val payload = JSONObject().apply {
                    put("reason", note)
                }

                OutputStreamWriter(conn.outputStream).use { it.write(payload.toString()) }

                if (conn.responseCode in 200..299) {
                    Result.success(true)
                } else {
                    Result.failure(Exception("Errore invio giustifica: HTTP ${conn.responseCode}"))
                }
            } catch (e: Exception) {
                Result.failure(e)
            }
        }

    override suspend fun bookColloquio(token: String, colloquioId: String): Result<Boolean> =
        withContext(Dispatchers.IO) {
            try {
                val url = URL("$baseUrl/colloqui/$colloquioId/book")
                val conn = (url.openConnection() as HttpURLConnection).apply {
                    requestMethod = "POST"
                    setRequestProperty("Authorization", "Bearer $token")
                    setRequestProperty("Content-Type", "application/json; charset=UTF-8")
                    doOutput = true
                    doInput = true
                    connectTimeout = 10000
                    readTimeout = 10000
                }

                if (conn.responseCode in 200..299) {
                    Result.success(true)
                } else {
                    Result.failure(Exception("Errore prenotazione colloquio: HTTP ${conn.responseCode}"))
                }
            } catch (e: Exception) {
                Result.failure(e)
            }
        }
}
